### PR TITLE
HSEARCH-3264 Search 6 groundwork - Add some basic documentation for 6.0.0.Alpha1

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net//browse/HSEARCH-3264

I ended up writing the whole "getting started" section. That should be enough for an Alpha1.
It's loosely based on the same section from Search 5, with adjustments to make it a bit simpler and cover both the Lucene and the Elasticsearch backend (which complicates things, unfortunately).

Run this to read the documentation in a more understandable format:

```
mvn clean install -pl documentation -am -DskipTests
xdg-open ./documentation/target/asciidoctor/en-US/html_single/index.html
```

### CODE REVIEW

File: documentation/src/main/asciidoc/analysis.asciidoc [hunk 1]
```diff
@@ -0,0 +1,36 @@
+[[analysis]]
+== Analysis
+// Search 5 anchors backward compatibility
+[[analyzer]]
+
+include::todo-placeholder.asciidoc[]
+
+////
+TODO The getting started section has a link pointing here and expects the section to
+include a detailed explanation of analysis, ,
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/analysis.asciidoc]: No, it's rubbish. Fixed.

File: documentation/src/main/asciidoc/analysis.asciidoc [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+To know which character filters, tokenizers and token filters are available,
+refer to the documentation specific to each backend:
+
+* For Lucene, either browse the Lucene JavaDoc or read the corresponding section on the
+link:http://wiki.apache.org/solr/AnalyzersTokenizersTokenFilters[Solr Wiki].
+* For Elasticsearch, have a look at the online documentation.
+If you want to use a built-in analyzer and not create your own:
+https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-analyzers.html[analyzers];
+if you want to define your own analyzer:
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/analysis.asciidoc]: My intent was to just provide links to lists of all available components; the link you gave explains how to create a custom analyzer, and I would expect our own documentation to explain that? Actually we already do it in the getting started, though not extensively.

File: documentation/src/main/asciidoc/architecture.asciidoc [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+== Architecture
+// Search 5 anchors backward compatibility
+[[search-architecture]]
+
+include::todo-placeholder.asciidoc[]
+
+////
+TODO The getting started section has a link pointing here and expects the section to
+include a description of each artifact,
+its Maven coordinates, and the corresponding path in the distribution bundle
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/architecture.asciidoc]: The architecture section is supposed to explain different architecture of a Hibernate-Search-based application, so I think listing all the components that can be used would make sense. But if you don't, I'm completely open to moving it to a different section. What's your take on this?

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 4]
```diff
@@ -1,37 +1,474 @@
...
+you cannot ignore bridges in the predicate DSL,
+and so on.
+* APIs are not considered stable yet.
+* The low-level parts of Hibernate Search 5 that made its efficiency haven't been
+ported to Hibernate Search 6 yet.
+This results in obvious performance bottlenecks,
+such as all the indexing works from a multi-threaded application being executed
+one after another and without any form of batching.
+
+The exact features that currently implemented and expected to work will be documented
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 5]
```diff
@@ -1,37 +1,474 @@
...
+** Mapping of entities using either annotation mapping or programmatic mapping.
+However, we do not offer all the options that were available in Search 5 (yet),
+and we do not support all the data types (yet).
+** Automatic indexing of ORM entities as they are persisted within a transaction
+(enabled automatically if Hibernate Search is in your classpath and the entity is mapped).
+** Search for ORM entities using their Hibernate Search index,
+and retrieving managed entities as results.
+However, we do not offer many options regarding how the entities are retrieved and initialized (yet).
+** Custom, user-defined type bridges, property bridges or value bridges.
+However, the API
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 6]
```diff
@@ -1,37 +1,474 @@
...
+|Hibernate ORM (for the ORM mapper)
+|Hibernate ORM *{hibernateVersion}*.
+|JPA (for the ORM mapper)
+|JPA *{jpaVersion}*.
+|===============
+
+=== Migration notes
+
+If you are upgrading an existing application from an earlier version of Hibernate Search to the latest release,
+make sure to check out the http://hibernate.org/search/documentation/migrate/[migration guide].
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: I intend to add it, or at least add a placeholder, when we release Alpha1.

I wanted to discuss it with you, though: what if, starting with Search 6, we moved the migration guides to a section of the user documentation, as suggested in [HSEARCH-3283](https://hibernate.atlassian.net/browse/HSEARCH-3283)? That would make sense in my opinion, and we could easily add a message on the website saying "for Hibernate Search 6 and above, see the "Migrating" section in the documentation". More importantly, we could, in the future, update the migration guide in the same PR where we add non-compatible changes...

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 7]
```diff
@@ -1,37 +1,474 @@
...
+
+[WARNING]
+====
+**To Hibernate Search 5 users**
+
+If you pull our artifacts from a Maven repository and you come from Hibernate Search 5,
+be aware that just bumping the version number will not be enough.
+
+In particular, the the group IDs changed from `org.hibernate` to `org.hibernate.search`,
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 8]
```diff
@@ -1,37 +1,474 @@
...
+most the artifact IDs changed to reflect the new mapper/backend design,
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: "most of the". Fixed.

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 9]
```diff
@@ -1,37 +1,474 @@
...
+(like moving away from using Lucene types in Hibernate Search APIs).
+====
+
+[[gettingstarted-dependencies]]
+=== Dependencies
+
+The Hibernate Search artifacts can be found in Maven's http://central.sonatype.org/[Central Repository]
+but are released first in the http://repository.jboss.org/nexus/content/groups/public-jboss/[JBoss Maven Repository].
+See also the https://community.jboss.org/wiki/MavenGettingStarted-Users[Maven Getting Started wiki page] to use
+the JBoss repository.
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: I copy/pasted that from Search 5, but yes I agree, it's not useful and may even be harmful. Let's add something to a separate section later if it's really necessary.
Fixed.

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 10]
```diff
@@ -1,37 +1,474 @@
...
+
+If you do not want to, or cannot, fetch the JARs from a Maven repository,
+you can get them from the
+link:http://sourceforge.net/projects/hibernate/files/hibernate-search/{hibernateSearchVersion}/[distribution bundle hosted at Sourceforge].
+
+In order to use Hibernate Search, you will need at least two direct dependencies:
+
+* a dependency to the "mapper", which extracts data from your domain model and maps it to indexable documents;
+* and a dependency to to the "backend", which allows to index and search these documents.
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 11]
```diff
@@ -1,37 +1,474 @@
...
+
+Below are are the most common setups and matching dependencies for a quick start;
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 12]
```diff
@@ -1,37 +1,474 @@
...
+<dependency>
+   <groupId>org.hibernate.search</groupId>
+   <artifactId>hibernate-search-backend-lucene</artifactId>
+   <version>{hibernateSearchVersion}</version>
+</dependency>
+----
++
+If you get Hibernate Search from the distribution bundle,
+copy the JARs from `dist/engine`, `dist/mapper/orm`, `dist/backend/lucene`,
+and their respective `lib` subdirectory.
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 13]
```diff
@@ -1,37 +1,474 @@
...
+<dependency>
+   <groupId>org.hibernate.search</groupId>
+   <artifactId>hibernate-search-backend-elasticsearch</artifactId>
+   <version>{hibernateSearchVersion}</version>
+</dependency>
+----
++
+If you get Hibernate Search from the distribution bundle,
+copy the JARs from `dist/engine`, `dist/mapper/orm`, `dist/backend/elasticsearch`,
+and their respective `lib` subdirectory.
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 14]
```diff
@@ -1,37 +1,474 @@
...
+[source, XML, indent=0]
+----
+include::{resourcesdir}/META-INF/persistence.xml[tags=gettingstarted-configuration-orm_lucene]
+----
+<1> Define a backend named "myBackend" relying on Lucene technology.
+<2> Define the storage for that backend as a local filesystem directory.
+<3> The backend will store indexes in the current working directory by default.
+If you want to store the indexes elsewhere,
+uncomment this line and set the value of the property.
+<4> Make sure to use the backend we just defined for all indexes
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 15]
```diff
@@ -1,37 +1,474 @@
...
+[source, XML, indent=0]
+----
+include::{resourcesdir}/META-INF/persistence.xml[tags=gettingstarted-configuration-orm_elasticsearch]
+----
+<1> Define a backend named "myBackend" relying on Elasticsearch technology.
+<2> The backend will attempt to connect to http://localhost:9200 by default.
+If you want to connect to another URL, uncomment these lines
+and set the value for the "host" property,
+and optionally the username and password.
+<3> Make sure to use the backend we just defined for all indexes
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 16]
```diff
@@ -1,37 +1,474 @@
...
 ----
 [source, JAVA, indent=0]
 ----
-include::{sourcedir}/org/hibernate/search/gettingstarted/model/after/Author.java[tags=author-entity-after-hsearch-gettingstarted]
+include::{sourcedir}/org/hibernate/search/documentation/gettingstarted/withhsearch/withoutanalysis/Author.java[tags=include;!getters-setters]
 ----
+<1> `@Indexed` marks `Book` as indexed, i.e. an index will be created for that entity, and that index will be kept up to date.
+<2> By default, the JPA `@Id` is used to generate a document identifier.
+<3> `@GenericField` maps a property to an index field with the same name and type.
+As such, the field is indexed in a way that only allow exact matches;
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 17]
```diff
@@ -1,37 +1,474 @@
...
+The following section will give a reasonable example of a general-purpose analyzer.
+For more advanced use cases, refer to the <<analysis>> section.
+
+==== Configuration
+
+Once you know what analysis is and which analyzer you want to apply,
+you will need to define it, or at least give it a name in Hibernate Search.
+This is done though analysis configurers, which are defined per backend:
+
+. First, you need implement an analysis configurer, a Java class that implements a backend-specific interface:
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/main/asciidoc/getting-started.asciidoc [hunk 18]
```diff
@@ -1,37 +1,474 @@
...
+`LuceneAnalysisConfigurer` or `ElasticsearchAnalysisConfigurer`.
+. Second, you need to alter the configuration of your backend to actually use your analysis configurer.
+
+As an example, let's assume that one of your indexed book entities has the title
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/main/asciidoc/getting-started.asciidoc]: Fixed

File: documentation/src/test/java/org/hibernate/search/documentation/gettingstarted/withhsearch/withoutanalysis/Author.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+import org.hibernate.search.mapper.pojo.mapping.definition.annotation.GenericField;
+
+@Entity // <5>
+public class Author {
+
+	@Id
+	@GeneratedValue
+	private Integer id;
+
+	@GenericField // <3>
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/test/java/org/hibernate/search/documentation/gettingstarted/withhsearch/withoutanalysis/Author.java]: In the getting started guide, we display the Book entity, then Author, in the same panel. Markers 1 through 4 appear in Book, 5 appears in Author, and then 3 appears again in Author. I could assign a new number and copy/paste the message from 3, but that would be even weirder, I think...

Is it weird if you look at the rendered version?

**Comment by gsmet** [comment]
Thanks, merged.

**Comment by romani** [comment]
@yrodiere , @gsmet , Travis was failed - https://travis-ci.com/hibernate/hibernate-search/builds/91700107#L741

and now it make problems for chekcstyle regression testing - https://github.com/checkstyle/checkstyle/pull/6215#issuecomment-439541183 and now all PRs are faliling, please fix checkstyle violation, or we will have to enforce validation of hibernate-search up to certain commit.

**Comment by yrodiere** [comment]
@romani Thanks for the heads-up! The checkstyle plugin was missing from that module configuration, and as a result checkstyle was only executed in the Travis build, where we execute checkstyle explicitly before we execute the full build. The Travis build being for contributors, the intent was to provide early feedback on common styling errors.

The Jenkins build that we rely on ourselves (in PRs, maintenance branches, ...) doesn't do that, so, with the missing plugin, checkstyle was not executed and the error didn't show up.

This should be fixed now.

**Comment by romani** [comment]
> doesn't do that, so, with the missing plugin, checkstyle was not executed and the error didn't show up.

@yrodiere ,
did you put checkstyle in Jenkins execution ?
if not please reference issue number when you plan to do this, to let us be safe on execution on hibernate-search HEAD. Unfortunately leak of violation could cause bunch of acton for checkstyle maintainers and all contributors who submitted PRs before.
Meanwhile we use strict commit state in checkstyle's CI to be safe and stable.

**Comment by yrodiere** [comment]
@romani Yes, checkstyle is executed as part of the main build in Jenkins. Believe me, it yells at me often enough :)

As I said, there was a configuration problem for a single module. The problem is now fixed on master (see https://github.com/hibernate/hibernate-search/commit/d263e9782f9fe49d4707ba5d1e7fbce07ea540be; the executions are configured through plugin management in the parent POM). So checkstyle will now be executed on all relevant modules both in Travis and in Jenkins.

**Pull request merged**